presto-orc/src/main/java/io/prestosql/orc/ValidationHash.java

```diff
@@ -0,0 +1,162 @@
+                    "rowHash",
+                    MethodType.methodType(long.class, Type.class, ValidationHash[].class, Block.class, int.class));
+            TIMESTAMP_HASH = lookup().findStatic(
+                    ValidationHash.class,
+                    "timestampHash",
+                    MethodType.methodType(long.class, Block.class, int.class));
+        }
+        catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public static ValidationHash createValidationHash(Type type)
+    {
+        requireNonNull(type, "type is null");
+        if (type.getTypeSignature().getBase().equals(MAP)) {
+            ValidationHash keyHash = createValidationHash(type.getTypeParameters().get(0));
+            ValidationHash valueHash = createValidationHash(type.getTypeParameters().get(1));
+            return new ValidationHash(MAP_HASH.bindTo(type).bindTo(keyHash).bindTo(valueHash));
+        }
+
+        if (type.getTypeSignature().getBase().equals(ARRAY)) {
+            ValidationHash elementHash = createValidationHash(type.getTypeParameters().get(0));
+            return new ValidationHash(ARRAY_HASH.bindTo(type).bindTo(elementHash));
+        }
+
+        if (type.getTypeSignature().getBase().equals(ROW)) {
+            ValidationHash[] fieldHashes = type.getTypeParameters().stream()
+                    .map(ValidationHash::createValidationHash)
+                    .toArray(ValidationHash[]::new);
+            return new ValidationHash(ROW_HASH.bindTo(type).bindTo(fieldHashes));
+        }
+
+        if (type.getTypeSignature().getBase().equals(StandardTypes.TIMESTAMP)) {
+            return new ValidationHash(TIMESTAMP_HASH);
+        }
+
+        return new ValidationHash(type.getHashCodeOperator(InvocationConvention.simpleConvention(FAIL_ON_NULL, BLOCK_POSITION)).getMethodHandle());
+    }
+
+    private final MethodHandle hashCodeOperator;
+
+    private ValidationHash(MethodHandle hashCodeOperator)
+    {
+        this.hashCodeOperator = requireNonNull(hashCodeOperator, "hashCodeOperator is null");
+    }
+
+    public long hash(Block block, int position)
+    {
+        if (block.isNull(position)) {
+            return NULL_HASH_CODE;
+        }
+        try {
+            return (long) hashCodeOperator.invokeExact(block, position);
+        }
+        catch (Throwable throwable) {
+            throwIfUnchecked(throwable);
+            throw new RuntimeException(throwable);
+        }
+    }
+
+    private static long mapSkipNullKeysHash(Type type, ValidationHash keyHash, ValidationHash valueHash, Block block, int position)
+    {
+        Block mapBlock = (Block) type.getObject(block, position);
+        long hash = 0;
+        for (int i = 0; i < mapBlock.getPositionCount(); i += 2) {
+            if (!mapBlock.isNull(i)) {
+                hash += keyHash.hash(mapBlock, i);
+                hash += valueHash.hash(mapBlock, i + 1);
+            }
+        }
```

Comment:
This will cause two maps with mixed keys and values to return the same hash:

{ 'x' -> 1, 'y' -> 2 } vs { 'x' -> 2, 'y' -> 1 }

Is that intentional?